lalegpl/lib/lale/nsga2.py

```diff
@@ -0,0 +1,567 @@
+        # Adjust max_evals if not a multiple of population size. This is
+        # required as Platypus performs evaluations in multiples of
+        # population_size.
+        adjusted_max_evals = (
+            self.max_evals // self.population_size
+        ) * self.population_size
+        if adjusted_max_evals != self.max_evals:
+            logger.info(
+                f"Adjusting max_evals to {adjusted_max_evals} from specified {self.max_evals}"
+            )
+
+        problem = Problem(parameter_num, target_num)
+        problem.types[:] = self.model_helper.types
+        problem.function = train_test_model
+
+        # Set the variator based on types of decision variables
+        varg = {}
+        first_type = problem.types[0].__class__
+        all_type_same = all([isinstance(t, first_type) for t in problem.types])
+        # use compound operator for mixed types
+        if not all_type_same:
+            varg["variator"] = CompoundOperator(SBX(), HUX(), PM(), BitFlip())
+
+        algorithm = NSGAII(
+            problem,
+            population_size=self.population_size,
+            **varg,
+        )
+
+        try:
+            algorithm.run(adjusted_max_evals, callback=time_check_callback)
+        except MaxBudgetExceededException as e:
+            logger.warning(
+                f"Max optimization time budget exceeded. Optimization exited prematurely.\n{e}"
+            )
+
+        solutions = nondominated(algorithm.result)
+        # solutions = [s for s in algorithm.result if s.feasible]`
+        # solutions = algorithm.result
+
+        moo_solutions = []
+        for solution in solutions:
+            vars = []
+            for pnum in range(parameter_num):
+                vars.append(problem.types[pnum].decode(solution.variables[pnum]))
+
+            vars_dict = self.param_to_dict(
+                vars,
+                self.model_helper.param_choices,
+                self.model_helper.param_categories,
+                self.model_helper.param_type,
+            )
+            moo_solutions.append(self.Soln(vars_dict, solution.objectives))
+            logger.info(f"{vars}, {solution.objectives}")
+
+        self.moo_solutions = moo_solutions
+
+        pareto_models = []
+        for solution in self.moo_solutions:
+            est = self.model_helper.create_instance(solution.variables)
+            est_trained = est.fit(X, y)
+            pareto_models.append((solution.variables, est_trained))
+
+        self.pareto_models = pareto_models
+        return self
+
+    def get_pareto_solutions(self):
+        return self.moo_solutions
+
+    def get_pareto_models(self):
+        return self.pareto_models
+
+    # Predict using first pareto-optimal estimator
+    def predict(self, X, **kwargs):
+        if "pipeline_name" in kwargs:
+            pname = kwargs["pipeline_name"]
+            pipeline = self.get_pipeline(pipeline_name=pname)
+            del kwargs["pipeline_name"]
```

Comment:
Why are we deleting something from an object sent by the user? Maybe this is something regarding python that i don't follow.

Reply:
Maybe use [dict_without](https://github.com/IBM/lale/blob/90c2348d0394a4cd7baaaeb4546f6589338ad382/lale/helpers.py#L145)?